packages/next/src/server-build.ts

```diff
@@ -1363,7 +1385,30 @@ export async function serverBuild({
       ...denormalizeNextDataRoute(),
 
       // /_next/data routes for getServerProps/getStaticProps pages
-      ...dataRoutes,
+      ...(isNextDataServerResolving
+        ? // when resolving data routes for middleware we need to include
+          // all dynamic routes including non-SSG/SSP so that the priority
+          // is correct
+          dynamicRoutes
+            .map(route => {
+              route = Object.assign({}, route);
+              route.src = path.posix.join(
+                '^/',
+                entryDirectory,
+                '_next/data/',
+                escapedBuildId,
+                route.src.replace(/(^\^|\$$)/g, '') + '.json$'
+              );
+
+              const { pathname } = new URL(route.dest || '/', 'http://n');
```

Comment:
Does this need to consider both absolute and relative urls for `dest`?

Reply:
These should only be relative paths as they need to match the related build output for a dynamic route